finder_server.cc

```diff
@@ -0,0 +1,237 @@
+/*
+ *
+ * Copyright 2015 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <iostream>
+#include <memory>
+#include <string>
+#include <cstdint>
+#include <utility>
+#include <vector>
+
+// #include <grpcpp/grpcpp.h>
+#include "finder_server.h"
+
+// #ifdef BAZEL_BUILD
+// #include "examples/protos/supplyfinder.grpc.pb.h"
+// #else
+// #include "supplyfinder.grpc.pb.h"
+// #endif
+
+using std::vector;
+using std::pair;
+using std::string;
+using grpc::Channel;
+using grpc::ClientContext;
+using grpc::ClientReader;
+using grpc::ClientReaderWriter;
+using grpc::ClientWriter;
+using grpc::Server;
+using grpc::ServerBuilder;
+using grpc::ServerContext;
+using grpc::ServerReader;
+using grpc::ServerReaderWriter;
+using grpc::ServerWriter;
+using grpc::Status;
+using grpc::StatusCode;
+using namespace supplyfinder;
+
+Status FinderServiceImpl::CheckFood (ServerContext* context, const Request* request,
+ ServerWriter<ShopInfo>* writer) {
+
+  std::cout << "========== Receving Request Food Id=" << request->food_id() << " ==========" << std::endl;
+
+  vector<ShopInfo> result = ProcessRequest(request->food_id());
```

Comment:
Again, you are writing a stream, but you are collecting the entire stream before sending.  Maybe you should send back results as you get them.

Reply:
I am collecting the entire stream because I want to sort by price and only return minimal information to the client. The finder service returns a minimal set of shops that satisfies the client's need.

Reply:
I collect the entire stream because I need to sort the list of shops by price and only return a minimal set of shops to the client.

Reply:
I think in that case this method should not be a streaming method.

Reply:
Resolved: Steaming method interleaves supplier requests and vendor requests, making the server more efficient